### ISSUE TITLE
Joins, merge, and has_many :through result in invalid table alias reference rejected by PG

### ISSUE THREAD

**Issue by nachokb** [rails/rails#19427]
ref https://gist.github.com/nachokb/a343b3f98f4d5d0bdca3

Important part: `Company.has_many :titles, through: :positions` and `Title.scope :x, -> (d) { joins(:positions).uniq.merge(Position.active_on(d)) }`

In this case, `Title.active_on(x)` generates a valid query, but `@company.titles` already adds a joins, and then `@company.titles.active_on(x)` adds a second one.

This second one uses an alias and somehow ends up first, like this: `INNER JOIN "positions" "positions_titles" ON "positions_titles"."title_id" = "titles"."id" LEFT OUTER JOIN "jobs" ON "jobs"."position_id" = "positions"."id"`, in which the `jobs` join references `positions.id` but positions is not yet defined so it's invalid.

Sqlite accepts it, but Postgresql doesn't.

**Comment by brainopia** [comment]
The reason for invalid query lies in `build_joins` (https://github.com/rails/rails/blob/8798788c65b0603caf4b55fe7e961e99f2ea4bde/activerecord/lib/active_record/relation/query_methods.rb#L944)

`join_list` will contain a link to `positions` table and `association_joins` will be `[:positions]`. `ActiveRecord::Associations::JoinDependency` constructs joins for associations, aliasing them if `join_list` contains links to the same tables. Therefore it will alias `:positions` association but `join_list` will be inserted only at the end https://github.com/rails/rails/blob/8798788c65b0603caf4b55fe7e961e99f2ea4bde/activerecord/lib/active_record/relation/query_methods.rb#L981.

So the alias is used because we have two joins to the same table. And a join with original table name will always be the last one. `join_infos` contains a join from `merge` using `positions` before it's defined at the end. A simple fix would be to move `manager.join_sources.concat(join_list)` before `join_infos`, and while it helps with the issue it breaks some other test in ActiveRecord.

**Comment by nachokb** [comment]
@brainopia thanks for your analysis, I'll try to take another look this week

**Comment by nachokb** [comment]
btw, I should update the gist to point to master...

**Comment by hadees** [comment]
I think this is [the same issue](https://github.com/devmynd/jsonb_accessor/issues/33) I saw on the gem `json_attributes`.

This code should be able to reproduce the error.

```
class Product < ActiveRecord::Base
  belongs_to :product_category
end

class ProductCategory < ActiveRecord::Base
  attribute "foo", ActiveRecord::Type::Value
  has_many :products
end

Product.includes(:product_category).order("product_categories.id").to_a
# raises:
# PG::UndefinedColumn: ERROR:  column product_categories.foo does not exist
#      LINE 1: ...S t1_r0, "product_categories"."options" AS t1_r1, "product_c...
#                                                                    ^
```

**Comment by nachokb** [comment]
Update: still happening on Rails 5.0.1 (updated Gist)

**Comment by niclin** [comment]
I have same issue.

**Comment by kamipo** [comment]
Fixed in #29413.
